common/flatpak-exports.c

```diff
@@ -105,8 +105,58 @@ struct _FlatpakExports
   GHashTable           *hash;
   FlatpakFilesystemMode host_etc;
   FlatpakFilesystemMode host_os;
+  int                   host_fd;
 };
 
+/*
+ * When populating /run/host, pretend @fd was the root of the host
+ * filesystem.
+ */
+void
+flatpak_exports_take_host_fd (FlatpakExports *exports,
+                              int fd)
+{
+  glnx_close_fd (&exports->host_fd);
+
+  if (fd >= 0)
+    exports->host_fd = fd;
+}
+
+static gboolean
+flatpak_exports_stat_in_host (FlatpakExports *exports,
+                              const char *abs_path,
+                              struct stat *buf,
+                              int flags,
+                              GError **error)
+{
+  g_return_val_if_fail (abs_path[0] == '/', FALSE);
+
+  if (exports->host_fd >= 0)
+    {
+      /* If abs_path is "/usr", then stat "usr" relative to host_fd.
+       * As a special case, if abs_path is "", stat host_fd itself. */
+      return glnx_fstatat (exports->host_fd, &abs_path[1], buf,
+                           AT_EMPTY_PATH | flags,
+                           error);
```

Comment:
This assumes that the `host_fd` does not point to a directory that contains inconvenient symlinks, for example `usr -> /really-usr`. That assumption is fine in practice for the test data generated by `test-exports` in this branch, and also fine for Flatpak's `/run/host`, but does not necessarily hold in general sysroots.

pressure-vessel has a general implementation of symlink resolution in a sysroot using `O_PATH` fds, with the correct fakechroot-like semantics where absolute symlinks are re-interpreted to be relative to the sysroot. I could copy that into Flatpak if you want (and we might need it anyway for #3797), but it seems like overkill for this particular use-case, so I'm inclined to omit it for now.

Reply:
Yeah, seems fine for now.